src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProviderHelper.cs

```diff
@@ -138,14 +139,19 @@ public static (AddRangeResult result, bool moreChildrenToRight) AddStorageRange(
 
             tree.UpdateRootHash();
 
-            if (tree.RootHash != expectedRootHash)
+            if (tree.RootHash.ValueHash256 != expectedRootHash)
             {
                 return (AddRangeResult.DifferentRootHash, true);
             }
 
-            StitchBoundaries(sortedBoundaryList, tree.TrieStore);
-
-            tree.Commit(writeFlags: WriteFlags.DisableWAL);
+            // This will work if all StorageRange requests share the same AccountWithPath object which seems to be the case.
+            // If this is not true, StorageRange request should be extended with a lock object.
+            // That lock object should be shared between all other StorageRange requests for same account.
+            lock (account.Account)
```

Comment:
I don't like the `account` based locking and using the object that might be elsewhere locked or used for different purposes. Locking should have specific scoping and should be reasoned about locally.

Another take is that this lock is needed, correct me @damian-orzechowski if I'm wrong, only if we know that there's no pending requests. We could extract the check from `StitchBoundaries` and assert it explicitly before the lock.

Reply:
we could create artificial object and pass it around, but I don't really see the point other than increasing the memory pressure and complicating the code. We added comments! :)

Reply:
Done some refactoring here #7930. The problem is that passing around the whole `StorageRange` request means also passing an index of the actual account processed (accounts are processed one at a time), so it doesn't look very nice.